### ISSUE TITLE
Add assertions while piping

### ISSUE THREAD

**Issue by gab23r** [pyjanitor-devs/pyjanitor#1128]
# Brief Description

A feature to be able to assert dataframe while piping

I would like to propose...

# Example API

```python
(
    df
    .assert_(lambda d: d.A.sum()==0)
    .pipe(...)
)

```

This feature is avaible in the package https://github.com/ClementWalter/pandas-addons
But I think that it is really missing in pyjanitor

**Comment by samukweku** [comment]
@gab23r kindly share an example, with a small dataframe, of what happens with `assert_`. The example should show what happens when it succeeds and when it does not

**Comment by ericmjl** [comment]
@gab23r thanks for chiming in! I think what you're looking for is available as the `.then()` or `.also()`. What these functions do is let you inject any arbitrary function into your dataframe piping chain, of which assertions are also possible. `.then()` expects you to explicitly return the dataframe while `.also()` will just execute your custom function while returning the original dataframe passed in.

That said, if you're doing assertions about the correctness of your dataframes, I'd suggest you check out `pandera` by @cosmicBboy. He and his team have done a ton of great work to make awesome tooling for validating pandas DataFrames; I use it in basically every project at work. You can find it here: https://github.com/unionai-oss/pandera

To summarize, I think you've got two options here:

1. Lightweight assertions using `.also()` and `lambda`s, or
2. Full-fledged dataframe validation using `pandera`.

Hope these answer your query! If you're amenable to it, please feel free to reply back or close the issue :smile:.

**Comment by cosmicBboy** [comment]
you could also pipe schemas :)

```python
import pandera as pa

schema = pa.DataFrameSchema({
    "A": pa.Column(checks=pa.Check(lambda A: A.sum() == 0)),
    "B": pa.Column(...),
    ...  # etc
})

(
    df
    .then(schema) # or you can use pipe
    .pipe(...)
)
```

Using pandera would make sense if you:
1. have logical groups of validation checks that are numerous and potentially complex
2. want to reuse these checks across your codebase (including your test suite)

As someone put it, pandera is like fancy assertions for dataframes.

**Comment by gab23r** [comment]
Thank you! I should definitely have a closer look to `pandera`

But for now, I would like to have a lightweight solution. However, how do you make the first solution working because assert is a statement so you can not do `lambda df: assert ...`. Do I miss something ?

Here is an working example of the desired feature.

```
# credit to https://github.com/ClementWalter/pandas-addons
def assert_(pandas_object, func, *args, error_type=AssertionError):
    """
    (
        pd.DataFrame({"a": [1, 2]})
        .assert_(lambda df: (df.a > 0).all())
        .assign(b=lambda df: 1 / df.a)
    )
    """
    try:
        cond = bool(func(pandas_object))
    except ValueError as original_error:
        raise ValueError(
            (
                "Cannot cast function output to bool, did you forget to add `.all()`, `.any()` or `.empty`?\n"
                "Original output:\n"
                f"{func(pandas_object)}"
            )
        ) from original_error
    if not cond:
        raise error_type(
            *[arg(pandas_object) if callable(arg) else arg for arg in args])
    return pandas_object

import pandas as pd
pd.DataFrame.assert_ = assert_


df = pd.DataFrame({
    'A' : [0, 1, 2],
    'B' : [0, 1, pd.NA]
})

(
    df
    .assert_(lambda d: not d.empty, 'DataFrame is empty')  # Ok
    .assert_(lambda d: d.notna().all().all(), 'DataFrame should not contain null values') # Fails !
)
```

**Comment by samukweku** [comment]
You can use pandas `pipe`:

```py
(df
.pipe(assert_, lambda d: not d.empty, 'DataFrame is empty') # passes
.pipe(assert_, lambda d: d.notna().all().all(), 'DataFrame should not contain null values') # fails
)
```

You can equally use janitor's `also` function:

```py
(
df
.also(assert_, lambda d: not d.empty, 'DataFrame is empty') # passes
.also(assert_, lambda d: d.notna().all().all(), 'DataFrame should not contain null values') # fails with error message
)
```

**Comment by ericmjl** [comment]
Hi @gab23r! I'm doing a bit of an issue cleanup here. Can I check if your question has been answered? If so, let us know, and we'll close the issue out.

**Comment by ericmjl** [comment]
@gab23r since there's been no movement on this issue, I'm going to close it. Please feel free to reopen, though!
